hazelcast/src/main/java/com/hazelcast/sql/impl/type/converter/DoubleConverter.java

```diff
@@ -38,27 +40,58 @@ private DoubleConverter() {
 
     @Override
     public byte asTinyint(Object val) {
-        return (byte) cast(val);
+        double casted = cast(val);
+        byte converted = (byte) casted;
+
+        if (converted != (int) casted || !Double.isFinite(casted)) {
+            throw cannotConvert(QueryDataTypeFamily.TINYINT, val);
+        }
+
+        return converted;
     }
 
     @Override
     public short asSmallint(Object val) {
-        return (short) cast(val);
+        double casted = cast(val);
+        short converted = (short) casted;
+
+        if (converted != (int) casted || !Double.isFinite(casted)) {
+            throw cannotConvert(QueryDataTypeFamily.SMALLINT, val);
+        }
+
+        return converted;
     }
 
     @Override
     public int asInt(Object val) {
-        return (int) cast(val);
+        double casted = cast(val);
+        int converted = (int) casted;
+
+        if (converted != (long) casted || !Double.isFinite(casted)) {
+            throw cannotConvert(QueryDataTypeFamily.INT, val);
+        }
+
+        return converted;
     }
 
     @Override
     public long asBigint(Object val) {
-        return (long) cast(val);
+        double casted = cast(val);
+        double truncated = casted > 0.0 ? Math.floor(casted) : Math.ceil(casted);
+        long converted = (long) truncated;
+
+        // No checks for NaNs and infinities are needed: NaNs are zeros and
+        // infinities are Long.MAX/MIN_VALUE when converted to long.
+        if ((double) converted != truncated) {
+            throw cannotConvert(QueryDataTypeFamily.BIGINT, val);
+        }
+
+        return converted;
     }
 
     @Override
     public BigDecimal asDecimal(Object val) {
-        return BigDecimal.valueOf(cast(val));
+        return new BigDecimal(Double.toString(cast(val)), DECIMAL_MATH_CONTEXT);
```

Comment:
Same concern as in the `FloatConverter`

Reply:
fixed in here https://github.com/hazelcast/hazelcast/pull/17183#discussion_r452036110